arroyo/processing/strategies/dead_letter_queue/policies/abstract.py

```diff
@@ -1,5 +1,19 @@
 from abc import ABC, abstractmethod
-from typing import Any, Sequence
+from dataclasses import dataclass
+from datetime import datetime
+from typing import Optional, Sequence, Union
+
+Serializable = Union[str, bytes]
+
+
+@dataclass(frozen=True)
+class InvalidMessage:
+    payload: Serializable
+    timestamp: datetime
+    reason: Optional[str] = None
+    consumer_group: Optional[str] = None
+    partition: Optional[int] = None
+    offset: Optional[int] = None
```

Comment:
No Kafka topic? I actually wonder if we should just remove all this and have the user of this library put it into the payload instead. Otherwise I feel like we will be changing this list too frequently as it is quite rigid and cannot account for everyone's usecase.

Reply:
The reason we dropped the Kafka topic is cause we decided each (consumer? storage?) will have its own dead letter topic. We don't need topic info here anymore since we know which one it is from which DLQ topic a message is produced to.

I also agree that we could just let the user decide entirely how the message should look, which is mostly why these are optional anyway.

Reply:
My question is really whether we should just remove all of these though? The inclusion of these particular items (and not topic for example) seems *extremely* specific to our usage

Reply:
The idea was to make it as simple as possible for the user to have the dead letter queue in place. If something is broken just raise a specific exception. 
I agree that these parameters are quite ad-hoc. 
What about this:
```
class InvalidMessage(ABC):
   @abtrsactmethod
   def to_dict() -> Mapping


@dataclass
class InvalidKafkaMessage(InvalidMessage):
   consumer_group
   offset 
   partition
   timestamp
   payload

   def to_dict() -> Mapping:
       impl

class InvalidRawMessage(InvalidMessage): 
   payload

  def to_dict() -> .....
```

So we have typed implementations and no ad-hoc optional parameters
The exception would contain a list of `InvalidMessage`

Reply:
I don't know if we need to add that much complexity since at the end the result will be the same in the DLQ topic. There isn't really a logical difference between the two dataclasses you suggest.

Eg.
```python3
InvalidRawMessage(payload=str({"consumer_group": "something", "offset": 2, ...}))
InvalidKafkaMessage(consumer_group="something", offset=2, ...)
```

Both of these will 'dictify' into the same thing pretty much, why not just have one class and let the user decide what they care about:
```python3
@dataclass
class InvalidMessage:
     payload: Serializable
```

Reply:
Eg. Metrics processors just have a `Mapping[str, Any]` as their message.

https://github.com/getsentry/snuba/blob/c568f271eaf5130bf790b2b4e7a429dcbaf6f259/snuba/datasets/metrics_bucket_processor.py#L33-L35

I'm sure that message is a `KafkaMessage` but they'd end up using `InvalidRawMessage` anyway for invalid messages.

Reply:
https://github.com/getsentry/arroyo/pull/64